rpcserver.go

```diff
@@ -3420,25 +3420,85 @@ func (r *rpcServer) GetNodeInfo(ctx context.Context,
 func (r *rpcServer) QueryRoutes(ctx context.Context,
 	in *lnrpc.QueryRoutesRequest) (*lnrpc.QueryRoutesResponse, error) {
 
-	// First parse the hex-encoded public key into a full public key object
-	// we can properly manipulate.
-	pubKeyBytes, err := hex.DecodeString(in.PubKey)
-	if err != nil {
-		return nil, err
+	// If a payment request was specified, extract the destination, amount,
+	// and hops from it.
+	payReq, err := zpay32.Decode(
+		in.PaymentRequest, activeNetParams.Params,
+	)
+	if err != nil && len(in.PaymentRequest) > 0 {
+		return nil, fmt.Errorf("failed to parse payment request")
 	}
-	pubKey, err := btcec.ParsePubKey(pubKeyBytes, btcec.S256())
-	if err != nil {
-		return nil, err
+
+	var (
+		pubKey     *btcec.PublicKey
+		amtMSat    lnwire.MilliSatoshi
+		routeHints [][]routing.HopHint
+	)
+
+	// Route hints can be read either from the invoice or from an
+	// explicit set of route hints provided.
+	if payReq != nil {
+		routeHints = payReq.RouteHints
+	}
+	// Explicit route hints take precedence (overrides) over hints
+	// provided in invoice
+	if len(in.Hints) > 0 {
+		for _, rh := range in.Hints {
+			var route []routing.HopHint
+			for _, h := range rh.HopHints {
+				var nodeID *btcec.PublicKey
+				if len(h.NodeId) > 0 {
+					nodeIDBytes, err := hex.DecodeString(h.NodeId)
+					if err != nil {
+						return nil, err
+					}
+					nodeID, err = btcec.ParsePubKey(nodeIDBytes, btcec.S256())
+					if err != nil {
+						return nil, err
+					}
+				}
+				route = append(route, routing.HopHint{
+					NodeID:                    nodeID,
+					ChannelID:                 h.ChanId,
+					FeeBaseMSat:               h.FeeBaseMsat,
+					FeeProportionalMillionths: h.FeeProportionalMillionths,
+					CLTVExpiryDelta:           uint16(h.CltvExpiryDelta),
+				})
+			}
+			routeHints = append(routeHints, route)
+		}
+	}
+
+	// Parse the hex-encoded public key into a full public key object
```

Comment:
you can do this check before parsing the hop hints?

Reply:
Yes. Moved up (need parsing of payment first though)